services/payments/statemachine_solana.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/blocto/solana-go-sdk/rpc"
+	"github.com/blocto/solana-go-sdk/types"
+	paymentLib "github.com/brave-intl/bat-go/libs/payments"
+	"github.com/mr-tron/base58"
+	"github.com/shopspring/decimal"
+)
+
+type chainIdempotencyData struct {
+	BlockHash   string            `json:"blockHash"`
+	SlotTarget  uint64            `json:"slotTarget"`
+	Transaction types.Transaction `json:"transaction"`
+}
+
+type TxnCommitmentStatus string
+
+const (
+	SPLBATMintDecimals   uint8               = 8                                              // Mint decimals for Wormhole wrapped BAT on mainnet
+	SPLBATMintAddress    string              = "EPeUFDgHRxs9xxEPVaL6kfGQvCon7jmAWKVUHuux1Tpz" // Mint address for Wormhole wrapped BAT on mainnet
+	TxnProcessed         TxnCommitmentStatus = "processed"
+	TxnConfirmed         TxnCommitmentStatus = "confirmed"
+	TxnFinalized         TxnCommitmentStatus = "finalized"
+	TxnNotFound          TxnCommitmentStatus = "notfound"
+	TxnUnknown           TxnCommitmentStatus = "unknown"
+	TxnInstructionFailed TxnCommitmentStatus = "instructionfailed"
+)
+
+// SolanaMachine is an implementation of TxStateMachine for Solana on-chain payouts
+// use-case.
+//
+// Including the baseStateMachine provides a default implementation of TxStateMachine.
+type SolanaMachine struct {
+	baseStateMachine
+	// signingKey is the private key of the funding wallet encoded in base58 format.
+	//
+	// The key can be derived from a mnemonic like this:
+	//
+	// 	mnemonic := "neither lonely flavor argue grass remind eye tag avocado spot unusual intact"
+	// 	seed := bip39.NewSeed(mnemonic, "") // (mnemonic, password)
+	// 	path := `m/44'/501'/0'/0'`
+	// 	derivedKey, _ := hdwallet.Derived(path, seed)
+	// 	derivedKey.PrivateKey
+	signingKey        string
+	solanaRpcEndpoint string
+	splMintAddress    string
+	splMintDecimals   uint8
+}
+
+func (sm *SolanaMachine) Authorize(ctx context.Context) (*paymentLib.AuthenticatedPaymentState, error) {
+	var err error
+	// Allow the base Authorize implementation to dictate error behavior until it succeeds
+	sm.transaction, err = sm.baseStateMachine.Authorize(ctx)
+	if err != nil {
+		return sm.transaction, err
+	}
+
+	// In the event of extra calls to Authorize (i.e. a third authorization when only two are needed)
+	// ExternalIdempotency will already be generated. Do not generate again. Do not error.
+	if sm.transaction.ExternalIdempotency != nil {
+		return sm.transaction, nil
+	}
+
+	// If the base Authorize method indicates we can proceed, generate, sign, and persist the
+	// transaction
+	client := solanaClient.NewClient(sm.solanaRpcEndpoint)
+	latestBlockhashResult, err := client.GetLatestBlockhashAndContext(ctx)
+	if err != nil {
+		return sm.transaction, fmt.Errorf("get recent block hash error, err: %v", err)
+	}
+	blockHash := latestBlockhashResult.Value.Blockhash
+	slotTarget := latestBlockhashResult.Context.Slot + 150
+
+	signer, err := types.AccountFromBase58(sm.signingKey)
```

Comment:
nit: this base58 encoding doesn't appear to be constant time. @evq are you aware of any go-lang constant time decoding libraries for base58? I did a quick look and didn't spot anything. FWIW, we didn't choose to use this but rather are picking up the dependency from the SDK we chose.

I'm not overly concerned with this because we're not doing with an oracle signing interaction. At worst, a malicious RPC service may be able to time when we submit transactions to the pool and try to conduct a timing attack. So, ideally we would be able to address encoding in constant time as well but I don't consider this a blocker.